### PR TITLE
(small breaking change) Merge v13's "TransactionNext" into the existing "Transaction" class

### PR DESCRIPTION
**Pull request by andreibancioiu**
The existing `Transaction` class has been adjusted to match the specs.

`TransactionNext` of v13 has been incorporated into `Transaction`.

## Breaking (but small) change on `tx.version` and `tx.options`

Before (v12):

```
// @deprecated
version: TransactionVersion;

// @deprecated
options: TransactionOptions;
```

Now (v13):

```
version: number;
options: number;
```

However, in practice, dApps should not be affected by this. The fields `version` and `options` were marked as deprecated for a long time. This change does NOT affect the following (since they rely on the not-changed `setVersion` and `setOptions`):
 - mx-sdk-js-hw-provider
 - sdk-dapp

There, version & options are mutated like this (example):

```
transaction.setVersion(TransactionVersion.withTxOptions());
transaction.setVersion(TRANSACTION_VERSION_WITH_OPTIONS);
transaction.setOptions(options | TRANSACTION_OPTIONS_TX_HASH_SIGN);
transaction.setOptions(TransactionOptions.withOptions({ guarded: true }));
transaction.setOptions(TransactionOptions.withOptions({
    guarded: true,
    ...(isLedger ? { hashSign: true } : {})
}));
```

These functions, `setVersion` and `setOptions`, remain available (legacy).

### CODE REVIEW

File: src/transaction.ts [hunk 1]
```diff
@@ -19,647 +31,599 @@ const TRANSACTION_HASH_LENGTH = 32;
...
+
+    /**
+     * The version, required by the Network in order to correctly interpret the contents of the transaction.
+     */
+    public version: number;
+
+    /**
+     * The options field, useful for describing different settings available for transactions.
+     */
+    public options: number;
```
Comment (andreibancioiu) [inline src/transaction.ts:92]: Breaking change.

File: src/transaction.ts [hunk 2]
```diff
@@ -19,647 +31,599 @@ const TRANSACTION_HASH_LENGTH = 32;
...
+        return feeForMove + processingFee;
     }
 
-    const gasPrice = transaction.gasPrice;
-    const feeForMove = moveBalanceGas * gasPrice;
-    if (moveBalanceGas === transaction.gasLimit) {
-        return feeForMove;
-    }
+    computeBytesForSigning(transaction: ITransaction): Uint8Array {
+        // TODO: do some checks for the transaction e.g. sender, chain ID etc.
```
Comment (andreibancioiu) [inline src/transaction.ts:567]: We should not forget about this.

File: src/transaction.ts [hunk 3]
```diff
@@ -19,647 +31,599 @@ const TRANSACTION_HASH_LENGTH = 32;
...
+        version,
+        options,
+        guardian,
+        signature,
+        guardianSignature,
+    }: {
+        nonce?: INonce | bigint;
+        value?: ITransactionValue | bigint;
+        sender?: IAddress | string;
+        receiver?: IAddress | string;
```
Comment (popenta) [inline src/transaction.ts:137]: shouldn't we have mandatory fields on the constructor? Like sender, receiver, gas limit and chain?

File: src/smartcontracts/smartContract.ts [hunk 4]
```diff
@@ -127,7 +127,7 @@ export class SmartContract implements ISmartContract {
         const bytecode = Buffer.from(code.toString(), 'hex');
         const metadataAsJson = this.getMetadataPropertiesAsObject(codeMetadata);
 
-        const nextTx = scNextTransactionFactory.createTransactionForDeploy({
+        const transaction = scNextTransactionFactory.createTransactionForDeploy({
```
Comment (popenta) [inline src/smartcontracts/smartContract.ts:130]: also rename `scNextTransactionFactory`

File: src/transaction.ts [hunk 5]
```diff
@@ -19,647 +31,599 @@ const TRANSACTION_HASH_LENGTH = 32;
...
+
+    /**
+     * The options field, useful for describing different settings available for transactions.
+     */
+    public options: number;
+
+    /**
+     * The address of the guardian.
+     */
+    public guardian: string;
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply src/transaction.ts:97]: Fixed.

File: src/transaction.ts [hunk 6]
```diff
@@ -19,647 +31,599 @@ const TRANSACTION_HASH_LENGTH = 32;
...
+
+    /**
+     * The address of the sender.
+     */
+    public sender: string;
+
+    /**
+     * The address of the receiver.
+     */
+    public receiver: string;
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply src/transaction.ts:52]: Fixed.

File: src/transaction.ts [hunk 7]
```diff
@@ -19,647 +31,599 @@ const TRANSACTION_HASH_LENGTH = 32;
...
+        gasPrice?: IGasPrice | bigint;
+        gasLimit?: IGasLimit | bigint;
+        data?: ITransactionPayload | Uint8Array;
+        chainID?: IChainID | string;
+        version?: ITransactionVersion | number;
+        options?: ITransactionOptions | number;
+        guardian?: IAddress | string;
+        signature?: Uint8Array;
+        guardianSignature?: Uint8Array;
+    }) {
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply src/transaction.ts]: Fixed.

File: src/transaction.ts (hunk 3 above)
Comment (andreibancioiu) [reply src/transaction.ts]: Indeed, fixed. Reverted, and now we have the same required fields as in PY (as in JS v12, as well): sender, receiver, chain ID and gas limit.

File: src/converters/transactionsConverter.ts [hunk 8]
```diff
@@ -0,0 +1,64 @@
+import { IPlainTransactionObject, ITransaction } from "../interface";
+import { Transaction } from "../transaction";
+
+export class TransactionsConverter {
```
Comment (danielailie) [inline src/converters/transactionsConverter.ts:4]: can you add public here?

**Review by danielailie — DISMISSED** [review]

**Review by danielailie — DISMISSED** [review]

**Review by radumojic — APPROVED** [review]

File: src/converters/transactionsConverter.ts [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+        return value && value.length ? Buffer.from(value).toString("base64") : undefined;
+    }
+
+    private toHexOrUndefined(value?: Uint8Array) {
+        return value && value.length ? Buffer.from(value).toString("hex") : undefined;
+    }
+
+    public plainObjectToTransaction(object: IPlainTransactionObject): Transaction {
+        const transaction = new Transaction({
+            nonce: Number(object.nonce),
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply src/converters/transactionsConverter.ts]: Indeed. Fixed.

**Review by popenta — APPROVED** [review]

**Review by radumojic — APPROVED** [review]

**Pull request merged**